File: SkyEditor.Core.Windows/Extensions/Online/OnlineExtensionCollection.vb
Imports System.Collections.Concurrent
Imports SkyEditor.Core.Extensions
Imports SkyEditor.Core.Utilities

Namespace Extensions.Online
    ''' <summary>
    ''' A collection of extensions stored online.
    ''' </summary>
    Public Class OnlineExtensionCollection
        Implements IExtensionCollection

        ''' <summary>
        ''' Creates a new instance of <see cref="OnlineExtensionCollection"/>.
        ''' </summary>
        ''' <param name="rootEndpoint">The root endpoint for connecting to the collection.</param>
        ''' <remarks>The api expects something like the following endpoints:
        ''' api/ExtensionCollection
        ''' api/ExtensionCollection/5
        ''' The endpoint name can vary, but the "/&lt;parentCollectionID&gt;" part must hold true.</remarks>
        Public Sub New(rootEndpoint As String)
            Client = New Net.WebClient
            rootEndpoint = rootEndpoint
            CachedInfo = New Dictionary(Of Integer, OnlineExtensionInfo)
        End Sub

        Public Sub New(rootEndpoint As String, parentCollectionId As Integer)
            Me.New(rootEndpoint)
            Me.ParentCollectionId = parentCollectionId
        End Sub

        Private Property Client As Net.WebClient
        Private Property RootEndpoint As String
        Private Property ParentCollectionId As Integer?
        Private Property GetExtensionsEndpoint As String
        Private Property CachedInfo As Dictionary(Of Integer, OnlineExtensionInfo)


        Private Async Function GetResponse() As Task(Of RootCollectionResponse)
            If _response Is Nothing Then
                Dim endpoint = RootEndpoint
                If ParentCollectionId.HasValue Then
                    endpoint &= "/" & ParentCollectionId.Value
                End If
                _response = Json.Deserialize(Of RootCollectionResponse)(Await Client.DownloadStringTaskAsync(New Uri(endpoint)).ConfigureAwait(False))
            End If
            Return _response
        End Function
        Private _response As RootCollectionResponse

        Public Async Function GetName() As Task(Of String) Implements IExtensionCollection.GetName
            If String.IsNullOrEmpty(_name) Then
                _name = (Await GetResponse.ConfigureAwait(False)).Name
            End If
            Return _name
        End Function
        Dim _name As String

        Public Async Function GetChildCollections(manager As PluginManager) As Task(Of IEnumerable(Of IExtensionCollection)) Implements IExtensionCollection.GetChildCollections
            If _childCollections Is Nothing Then
                For Each item In (Await GetResponse.ConfigureAwait(False)).ChildCollections
                    _childCollections.Add(New OnlineExtensionCollection(Me.RootEndpoint, item.ID))
                Next
            End If
            Return _childCollections
        End Function
        Dim _childCollections As List(Of OnlineExtensionCollection)

        Public Async Function GetExtensionCount(manager As PluginManager) As Task(Of Integer) Implements IExtensionCollection.GetExtensionCount
            If Not _extensionCount.HasValue Then
                _extensionCount = (Await GetResponse.ConfigureAwait(False)).ExtensionCount
            End If
            Return _extensionCount
        End Function
        Dim _extensionCount As Integer?

        Public Async Function GetExtensions(skip As Integer, take As Integer, manager As PluginManager) As Task(Of IEnumerable(Of ExtensionInfo)) Implements IExtensionCollection.GetExtensions
            Dim responseRaw = Await Client.DownloadStringTaskAsync((Await GetResponse.ConfigureAwait(False)).GetExtensionListEndpoint & $"/{skip}/{take}")
            Dim response = Json.Deserialize(Of List(Of OnlineExtensionInfo))(responseRaw)

            Dim i As Integer = skip
            For Each item In response
                item.IsInstalled = ExtensionHelper.IsExtensionInstalled(item, manager)

                If CachedInfo.ContainsKey(i) Then
                    CachedInfo(i) = item
                Else
                    CachedInfo.Add(i, item)
                End If
                i += 1
            Next

            Return response
        End Function

        Public Async Function InstallExtension(extensionID As String, version As String, manager As PluginManager) As Task(Of ExtensionInstallResult) Implements IExtensionCollection.InstallExtension
            'Download zip
            Dim tempName = manager.CurrentIOProvider.GetTempFilename
            Await Client.DownloadFileTaskAsync((Await GetResponse.ConfigureAwait(False)).DownloadExtensionEndpoint & $"/{extensionID}/{version}", tempName)

            'Install
            Dim result = Await ExtensionHelper.InstallExtensionZip(tempName, manager).ConfigureAwait(False)

            'Clean up
            manager.CurrentIOProvider.DeleteFile(tempName)

            Return result
        End Function

        Public Function UninstallExtension(extensionID As String, manager As PluginManager) As Task(Of ExtensionUninstallResult) Implements IExtensionCollection.UninstallExtension
            Dim typeName As String = (From c In CachedInfo.Values Where c.ID = extensionID Select c.ExtensionTypeName).First
            Return ExtensionHelper.UninstallExtension(typeName, extensionID, manager)
        End Function
    End Class
End Namespace

File: SkyEditor.Core/Extensions/Online/OnlineExtensionCollection.cs
using SkyEditor.Core.Utilities;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace SkyEditor.Core.Extensions.Online
{
    /// <summary>
    /// A collection of extensions stored online.
    /// </summary>
    public class OnlineExtensionCollection : IExtensionCollection
    {

        /// <summary>
        /// Creates a new instance of <see cref="OnlineExtensionCollection"/>.
        /// </summary>
        /// <param name="rootEndpoint">The root endpoint for connecting to the collection.</param>
        /// <remarks>The api expects something like the following endpoints:
        /// api/ExtensionCollection
        /// api/ExtensionCollection/5
        /// The endpoint name can vary, but the "/&lt;parentCollectionID&gt;" part must hold true.</remarks>
        public OnlineExtensionCollection(string rootEndpoint)
        {
            this.Client = new HttpClient();
            this.RootEndpoint = rootEndpoint;
            this.CachedInfo = new Dictionary<int, OnlineExtensionInfo>();
        }

        public OnlineExtensionCollection(string rootEndpoint, string parentCollectionId) : this(rootEndpoint)
        {
            this.ParentCollectionId = parentCollectionId;
        }

        private HttpClient Client { get; set; }
        private string RootEndpoint { get; set; }
        private string ParentCollectionId { get; set; }
        private string GetExtensionsEndpoint { get; set; }
        private Dictionary<int, OnlineExtensionInfo> CachedInfo { get; set; }

        private async Task<RootCollectionResponse> GetResponse()
        {
            if (ReferenceEquals(_response, null))
            {
                var endpoint = RootEndpoint;
                if (!string.IsNullOrEmpty(ParentCollectionId))
                {
                    endpoint += "/" + ParentCollectionId;
                }
                _response = Json.Deserialize<RootCollectionResponse>(await Client.GetStringAsync(new Uri(endpoint)).ConfigureAwait(false));
            }
            return _response;
        }
        private RootCollectionResponse _response;

        public async Task<string> GetName()
        {
            if (string.IsNullOrEmpty(_name))
            {
                _name = System.Convert.ToString((await GetResponse().ConfigureAwait(false)).Name);
            }
            return _name;
        }
        string _name;

        public async Task<IEnumerable<IExtensionCollection>> GetChildCollections(PluginManager manager)
        {
            if (ReferenceEquals(_childCollections, null))
            {
                _childCollections = new List<OnlineExtensionCollection>();
                foreach (var item in (await GetResponse().ConfigureAwait(false)).ChildCollections)
                {
                    _childCollections.Add(new OnlineExtensionCollection(this.RootEndpoint, item.ID));
                }
            }
            return _childCollections;
        }
        List<OnlineExtensionCollection> _childCollections;

        public async Task<int> GetExtensionCount(PluginManager manager)
        {
            if (!_extensionCount.HasValue)
            {
                _extensionCount = (await GetResponse().ConfigureAwait(false)).ExtensionCount;
            }
            return _extensionCount.Value;
        }
        int? _extensionCount;

        public async Task<IEnumerable<ExtensionInfo>> GetExtensions(int skip, int take, PluginManager manager)
        {
            var responseRaw = await Client.GetStringAsync((await GetResponse().ConfigureAwait(false)).GetExtensionListEndpoint + $"/{skip}/{take}");
            var response = Json.Deserialize<List<OnlineExtensionInfo>>(responseRaw);

            int i = skip;
            foreach (var item in response)
            {
                item.IsInstalled = ExtensionHelper.IsExtensionInstalled(item, manager);

                if (CachedInfo.ContainsKey(i))
                {
                    CachedInfo[i] = item;
                }
                else
                {
                    CachedInfo.Add(i, item);
                }
                i++;
            }

            return response;
        }

        public async Task<ExtensionInstallResult> InstallExtension(string extensionID, string version, PluginManager manager)
        {
            //Download zip
            var tempName = manager.CurrentIOProvider.GetTempFilename();
            manager.CurrentIOProvider.WriteAllText(tempName, await Client.GetStringAsync((await GetResponse().ConfigureAwait(false)).DownloadExtensionEndpoint + $"/{extensionID}/{version}"));

            //Install
            var result = await ExtensionHelper.InstallExtensionZip(tempName, manager).ConfigureAwait(false);

            //Clean up
            manager.CurrentIOProvider.DeleteFile(tempName);

            return result;
        }

        public Task<ExtensionUninstallResult> UninstallExtension(string extensionID, PluginManager manager)
        {
            string typeName = CachedInfo.Values.Where(x => x.ID == extensionID).Select(x => x.ExtensionTypeName).First();
            return ExtensionHelper.UninstallExtension(typeName, extensionID, manager);
        }
    }
}
